### PR TITLE
storage: Add a TimestampCache minimum size for eviction threshold

### PR DESCRIPTION
**Pull request by nvb**
As discussed, a minimum ts cache size before beginning evictions will
avoid transactions driven by an interactive shell from being aborted
after only 10 seconds. The threshold size of 512 is arbitrary, but it
sounds like a good number for now.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/5526)

### CODE REVIEW

File: storage/timestamp_cache.go [hunk 1]
```diff
@@ -36,6 +36,24 @@ const (
 	MinTSCacheWindow = 10 * time.Second
 )
 
+var (
+	// evictionSizeThreshold is the maximum allowed number of intervals
```
Comment (spencerkimball) [inline storage/timestamp_cache.go]: s/maximum/minimum/

File: storage/timestamp_cache.go [hunk 2]
```diff
@@ -36,6 +36,24 @@ const (
...
+	// in the TimestampCache before it will begin evicting intervals
+	// based on the time window. This threshold is intended to permit
+	// transactions to take longer than the evition window duration
+	// if the size of the cache is not a concern, such as when a user
+	// is using an interactive SQL shell.
+	evictionSizeThreshold = 512
+)
+
+func removeEvictionSizeThreshold() func() {
```
Comment (spencerkimball) [inline storage/timestamp_cache.go]: How about rename this `s/remove/disable/`?

**Comment by spencerkimball** [comment]
LGTM

File: storage/timestamp_cache.go [hunk 3]
```diff
@@ -36,6 +36,24 @@ const (
...
 )
 
+var (
+	// evictionSizeThreshold is the maximum allowed number of intervals
+	// in the TimestampCache before it will begin evicting intervals
+	// based on the time window. This threshold is intended to permit
+	// transactions to take longer than the evition window duration
+	// if the size of the cache is not a concern, such as when a user
+	// is using an interactive SQL shell.
+	evictionSizeThreshold = 512
```
Comment (tschottdorf) [inline storage/timestamp_cache.go]: do we really need to throw around globals for this? Just make it a setting on the timestamp cache.

File: storage/timestamp_cache.go [hunk 4]
```diff
@@ -36,6 +36,24 @@ const (
...
+)
+
+func removeEvictionSizeThreshold() func() {
+	orig := evictionSizeThreshold
+	evictionSizeThreshold = -1
+	return func() {
+		evictionSizeThreshold = orig
```
Comment (tschottdorf) [inline storage/timestamp_cache.go]: this would just go away.

**Comment by RaduBerinde** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/timestamp_cache.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-KDarPURbni-C6IAM56W-r1-40:-KDdH6E89THFPQq3iqSt:-1982632169) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b6ac5a439bd6736acc3330d777ce59fa357396e/storage/timestamp_cache.go#L40)):*
i understand how both `maximum` and `minimum` kind of make sense depending how you think about it :) 

Maybe `evictionSizeThreshold allows old entries to stay in the TimestampCache indefinitely as long as the number of intervals in the cache doesn't exceed this value. Once the cache grows beyond it, intervals are evicted according to the time window. This threshold is intended to ..`

---

*[storage/timestamp_cache.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-KDdGzYuPe91A4KRiZul:-KDdGzYuPe91A4KRiZum:-1068880788) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b6ac5a439bd6736acc3330d777ce59fa357396e/storage/timestamp_cache.go#L43)):*
eviction window

---

*[storage/timestamp_cache_test.go, line 181 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-KDdIZMQ3pBbFVUF9kMi:-KDdIZMQ3pBbFVUF9kMj:-278259612) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b6ac5a439bd6736acc3330d777ce59fa357396e/storage/timestamp_cache_test.go#L181)):*
We could consolidate these two - start by verifying intervals are not evicted like in the new test, then add some dummy intervals until the `evictionSizeThreshold` limit is reached, then verify we do evict intervals as expected. We wouldn't need to muck with `evictionSizeThreshold`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526)*

**Comment by RaduBerinde** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-:-KDdIvF27wXiaJGwbo0f:1417088721)*

**Comment by nvanbenschoten** [comment]
TFTR

---

Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/timestamp_cache.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-KDarPURbni-C6IAM56W-r1-40:-KDdv-IdnlIgFmtDaAzu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b6ac5a439bd6736acc3330d777ce59fa357396e/storage/timestamp_cache.go#L40)):*
Done.

---

*[storage/timestamp_cache.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-KDdGzYuPe91A4KRiZul:-KDdv1jYhZLL8QCwSQ--:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b6ac5a439bd6736acc3330d777ce59fa357396e/storage/timestamp_cache.go#L43)):*
Done.

---

*[storage/timestamp_cache.go, line 46 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-KDarPURbni-C6IAM56W-r1-46:-KDdv2NBRxBM7U8jMfi2:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b6ac5a439bd6736acc3330d777ce59fa357396e/storage/timestamp_cache.go#L46)):*
Done.

---

*[storage/timestamp_cache_test.go, line 181 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-KDdIZMQ3pBbFVUF9kMi:-KDdx6pnVFLELcLH98HP:260066716) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b6ac5a439bd6736acc3330d777ce59fa357396e/storage/timestamp_cache_test.go#L181)):*
Now that `evictionSizeThreshold` is a member on the cache, it's easy enough just to set it to zero for a test.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-:-KDdxtbaTbYeXa2UIR_c:677071605)*

**Comment by RaduBerinde** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5526#-:-KDe02vRFY6s1GMZ0_5-:1417088721)*

**Pull request merged**